### PR TITLE
Call memcpy to copy a struct

### PR DESCRIPTION
**Pull request by htibosch**
I have never trusted compilers when copying a struct in an assignment:
~~~
	struct xStruct { int a, b, c; };
	xStruct s1, s2;
	s1 = s2;
~~~

In [one occasion](https://github.com/FreeRTOS/Lab-Project-FreeRTOS-FAT/blob/5760fc95ab0f7a7d86f4fcf16b8e9ab5b7aec03a/ff_stdio.c#L1492) in FreeRTOS+FAT, there is such an assignment:
~~~diff
-    pxFindData->xDirectoryEntry.xModifiedTime = pxFindData->xDirectoryEntry.xCreateTime;        /* Date and Time Modified. */
+    /* Date and Time Modified. */
+    memcpy( &( pxFindData->xDirectoryEntry.xModifiedTime ),
+            &( pxFindData->xDirectoryEntry.xCreateTime ),
+            sizeof pxFindData->xDirectoryEntry.xModifiedTime );
-    pxFindData->xDirectoryEntry.xAccessedTime = pxFindData->xDirectoryEntry.xCreateTime;        /* Date of Last Access. */
+    /* Date of Last Access. */
+    memcpy( &( pxFindData->xDirectoryEntry.xAccessedTime ),
+            &( pxFindData->xDirectoryEntry.xCreateTime ),
+            sizeof pxFindData->xDirectoryEntry.xAccessedTime );
~~~
The above change helps avoiding a fatal exception. I've never seen this crashing before. I am compiling it for a STM32F7xx.

### CODE REVIEW

**Comment by htibosch** [comment]
I just checked the cause of the crash:

The parent struct `FF_FindData_t` is either declared on stack, or allocated by `pvPortKalloc()`. The start address is always 32-bit aligned.

However, the time structs can be 16-bit aligned because they consist of `uint16_t` members only:
~~~c
typedef struct
{
    uint16_t Year;   /* Year	(e.g. 2009). */
    uint16_t Month;  /* Month	(e.g. 1 = Jan, 12 = Dec). */
    uint16_t Day;    /* Day	(1 - 31). */
    uint16_t Hour;   /* Hour	(0 - 23). */
    uint16_t Minute; /* Min	(0 - 59). */
    uint16_t Second; /* Second	(0 - 59). */
} FF_SystemTime_t;
~~~
So far the compiler is correct. However the same compiler assumes that all structs are 32-bit aligned when doing an implicit `memcpy()`.
I will add a 2 filler bytes to make the structs 32-bit aligned, which is more efficient for `memcpy()`.

File: ff_stdio.c [hunk 1]
```diff
@@ -1489,8 +1489,14 @@ int ff_findnext( FF_FindData_t * pxFindData )
                     pxFindData->xDirectoryEntry.xCreateTime.Hour = ( uint16_t ) xTimeStruct.tm_hour;            /* Hour (0 - 23). */
                     pxFindData->xDirectoryEntry.xCreateTime.Minute = ( uint16_t ) xTimeStruct.tm_min;           /* Min (0 - 59). */
                     pxFindData->xDirectoryEntry.xCreateTime.Second = ( uint16_t ) xTimeStruct.tm_sec;           /* Second (0 - 59). */
-                    pxFindData->xDirectoryEntry.xModifiedTime = pxFindData->xDirectoryEntry.xCreateTime;        /* Date and Time Modified. */
-                    pxFindData->xDirectoryEntry.xAccessedTime = pxFindData->xDirectoryEntry.xCreateTime;        /* Date of Last Access. */
+                    /* Date and Time Modified. */
+                    memcpy( &( pxFindData->xDirectoryEntry.xModifiedTime ),
+                            &( pxFindData->xDirectoryEntry.xCreateTime ),
+                            sizeof pxFindData->xDirectoryEntry.xModifiedTime );
```
Comment (AniruddhaKanhere) [inline ff_stdio.c:1495]:
```suggestion
                            sizeof( pxFindData->xDirectoryEntry.xModifiedTime ) );
```

File: ff_stdio.c [hunk 2]
```diff
@@ -1489,8 +1489,14 @@ int ff_findnext( FF_FindData_t * pxFindData )
...
+                    /* Date of Last Access. */
+                    memcpy( &( pxFindData->xDirectoryEntry.xAccessedTime ),
+                            &( pxFindData->xDirectoryEntry.xCreateTime ),
+                            sizeof pxFindData->xDirectoryEntry.xAccessedTime );
```
Comment (AniruddhaKanhere) [inline ff_stdio.c:1499]:
```suggestion
                            sizeof( pxFindData->xDirectoryEntry.xAccessedTime ) );
```

File: include/ff_dir.h [hunk 3]
```diff
@@ -64,6 +64,8 @@ typedef struct
     uint32_t ulAddrCurrentCluster;
     uint32_t ulDirCluster;
     uint16_t usCurrentItem;
+    /* Insert 2 bytes to make the next struct 32-bit aligned. */
+    uint16_t usFillerBytes;
```
Comment (AniruddhaKanhere) [inline include/ff_dir.h:68]: I assume that there will be very few instances of this struct. And we'll not waste a lot of memory with filler bytes.

**Review by AniruddhaKanhere — COMMENTED** [review]
Thanks @htibosch for this PR.

I have a couple of suggestions and one question. Can you please take a look?

Thanks

File: include/ff_dir.h (hunk 3 above)
Comment (htibosch) [reply include/ff_dir.h:68]: Yes indeed, there will be very few instances of this struct. It is used to list files and directories.

**Review by htibosch — COMMENTED** [review]
Thanks for reviewing!

**Pull request merged**